feat(opinion_page): add a button to access RECAP doc attachments

florean opened the pull request:
After thinking for a long while about issue #5770, I think I've come up with a good solution.  This modifies the RECAP document page to display a button on attachments or documents with attachments which links to the next attachment associated with the document or the original document if viewing the last attachment, with a pulldown menu that allows direct access to every attachment.  This enables users to interact with the full docket entry without having to go back to the docket.  And it does it without any additional hits to the database.
<img width="593" alt="image" src="https://github.com/user-attachments/assets/d888f0ec-6786-43a5-ad7c-a79885f6dd4e" />
For the UI, I've made what I think are the best choices, but it is subjective and it's ultimately up to your taste.  Some things I struggled with: is "Attachment/Document #" clearer than using a truncated description?  Should the description length be more or less than 30 characters?  Are carets the best way to highlight the current doc in the dropdown?  Should the current doc be hyperlinked in the pulldown?  I've made my choices (no, no, yes, yes), but I can't say they're right for CourtListener.

Testing:  I've done a lot of manual testing while iterating over different designs, but I have limited and imperfect data.  I have covered documents with 0-3 attachments as well as documents and attachments with and without descriptions.  It'd be nice to see it in a staging environment with the documents that have the most attachments, but I think the design will hold up.

Resolves #5770

mlissner commented:
Thanks for playing with this. I think you've got an OK direction, but I find it confusing that other documents would be listed as actions for the one I'm currently on. 

What I think might work better could be integrating this into the heading: 

![image](https://github.com/user-attachments/assets/5d04e302-2095-4952-a1b6-4eb27a276c31)

For example, what if that has a drop down integrated into it where you could select the other items instead of #2-1? 

For testing, I don't know the item with the most attachments, but it's probably somewhere around 100 or something like that.

florean commented:
So I can see how it would work in your screenshot if clicking on `Attachment #1` popped up a dropdown to select other attachments.  But how do you get there from the root document?  What does the header look like for a document that has no attachments and a document that has attachments?

mlissner commented:
Maybe we do a really simple thing and just have a dropdown at the end of that heading. It'd look roughly like this: 

![image](https://github.com/user-attachments/assets/c19eb71f-0d92-4935-a34b-7dddc667d9ba)

In the menu, you could have something like: 

```
Main Document
Att. 1 - Exhibit A
Att. 2 - Civil Cover Sheet
...
```

Notes:
 - I'm not sure if a bullet (`•`), whitespace, a dash or an em-dash is best as the divider in there.
 - I don't think it needs to indicate which you're on. You know that from the header already.
 - I think if we do this, we can make it so that the `#1` in this heading isn't a link anymore, since this is a more useful approach.

What do you think? It's less discoverable, but advanced users will find it.

florean commented:
Ah, gotcha, I think that clears it up for me.  I'll take a pass at it.

florean commented:
Added a shorter tooltip to help with longer titles.
<img width="805" alt="image" src="https://github.com/user-attachments/assets/ef8d1910-d571-4630-9c17-79de5998a3aa" />

florean commented:
Since I don't think anyone has really reviewed the code yet, do you might if I do a rebase?  Half the commits at this point are merges from `main` and I really don't want to contribute to the 25% of `main` commits that are just merges from `main`.

> I have some concerns about how it might perform and about its UX once it goes live, but let's give it a go and see how it works.

Can you elucidate?  Are you concerned about the menu items?  Would you prefer a `{{ doc.document_number }} – {{ doc.description }}` style?  I kind of liked the consistency with the heading, but that format puts the most important information at the beginning (plus I can move the heading style to the tooltip).  It's really a 50/50 call to me and it's your site.  I'll be a little surprised if you get critical feedback once it goes live, so now's the best time to make any tweaks or changes you want.  At the very least, I can fork this branch and you can compare the two to see what feels best.  (If you don't have a staging server, maybe we could hide both behind Waffle flags and merge them after review?  Not sure how difficult creating new flags is, but it'd be an easy change to `view_recap_document`.)

I'm a little proud of how this turned out, especially being able to do it while increasing performance, and I'd like it to be the best it can be.  I'm new to your design and taste (as evidenced by the initial button) and I'm open to feedback.

Cheers!

P.S.  On, and I'll add a test case for `view_recap_document` before merge.

florean commented:
BTW, I sent a Contact Us message asking for some API token privileges to help with testing.  Posting here to maybe bump it up the queue.  Thanks!

florean commented:
Alright, I'll address any review feedback and leave this in as the default, but put an alternate behind a Waffle flag.  We'll tweak that once it's live and then integrate the improvements.

Thanks for the API access.  I got some good data from it and it was fortuitous, because I'd just accidentally nuked all of my local data with a Docker build.

elisa-a-v commented:
Nice, thank you! I'm marking as draft in the meantime, can you mark as ready for review once all tests are pushed please?

Let me know if there's anything you need feedback on before that anyway :)

florean commented:
Sorry, took a little longer than expected.

florean commented:
I'm sure you're busy and it probably doesn't make sense to look at this until you can review the whole thing, but the two things I'm concerned about are the custom CSS and the views code.

Thanks!

elisa-a-v reviewed (changes requested):
First of all, thank you for the PR and for your patience with the review 🙏🏽 

This is a nice feature that I'm sure will boost UX when navigating through a doc's attachments. I really like what you did with the dropdown design, I fully agree with you that highlighting the current item is better for UX, and CSS looks good to me save for one issue described below.

## Observations from testing

### 1. Docs with many attachments

#### Page overflow

For documents with a common number of attachments the dropdown looks great, but when the attachment list grows, the dropdown grows indefinitely too.

For example I found [this item](https://www.courtlistener.com/docket/17366820/akai-custom-guns-llc-v-kkm-precision-inc/?entry_gte=207#entry-207) with 212 attachments, and it overflows the page:

<details>
<summary>Overflow screenshot</summary>

<img width="400" alt="image" src="https://github.com/user-attachments/assets/2e316ea3-85d8-4e51-96cb-4ce9942b2967" />

</details>

Adding a `max-height` and `overflow: auto` to the dropdown should probably do the trick.

#### Performance considerations

I noticed the number of DB queries is reduced and docs with fewer attachments are indeed quite quicker to load, however for docs with too many attachments the time it takes the DB to perform those queries is considerably longer than what it was before. Hiding the dropdown for docs with too many attachments seems like the easy way out here, but we may want to wait to see how it performs in prod before doing anything about this. I tested with the dev DB which is not all that powerful so this may not be an issue in prod.

#### Tip on data cloning for local testing

> maybe if I get better at automatically generating some data I'll take a look at it.

I think you should be able to clone that docket with all its entries by running this command:

```
manage.py clone_from_cl --type search.Docket --id 17366820 --add-docket-entries
```

You can read more about it in [the command docstring](https://github.com/freelawproject/courtlistener/blob/main/cl/scrapers/management/commands/clone_from_cl.py), let me know if you need help with this.

### 2. Other RECAPDocument tabs

I only see the dropdown button when in the default PDF/Text pages, but it disappears when in the authorities tab:

<details>
<summary>PDF/Text tabs screenshot</summary>

<img width="400" alt="Image" src="https://github.com/user-attachments/assets/0e73230a-7c56-4a53-8545-fcb55a9cc6ce" />

</details>

<details>
<summary>Authorities tabs screenshot</summary>

<img width="400" alt="Image" src="https://github.com/user-attachments/assets/1b028243-4d98-45a9-98a9-5fee25980fff" />

</details>

The bug persists when changing tabs after navigating to the authorities tab. I noticed the URL doesn't go back to the main doc page but remains in the `/autorities/` url, I'm guessing these issues are related.

---

Great job overall, thank you for your thoughtful contributions! We'll be ready to merge once we've addressed the observations above and settled on the questions I left inline.

mlissner commented on cl/search/models.py:
```diff
@@ -1546,9 +1546,8 @@ def save(
-                else:
-                    # Only one duplicate. Attempt auto-resolution.
-                    other = others[0]
+                # Only one duplicate. Attempt auto-resolution.
+                other = others[0]
```
Tweak is fine. We check if others exist, and they do, so we check if there are more than 1 other. If so, we abort with a ValidationError. So now we know others is greater than zero (they exist, and not greater than 1). That can only mean one thing, that there's only one duplicate and we can attempt auto-resolution.

I can't see how it could be wrong.

mlissner commented on cl/opinion_page/views.py:
```diff
@@ -541,10 +541,10 @@ async def docket_authorities(
-    de = await DocketEntry.objects.aget(id=rd.docket_entry_id)
-    d = await Docket.objects.aget(id=de.docket_id)
-    court = await Court.objects.aget(id=d.court_id)
+def make_rd_title(rd: RECAPDocument) -> str:
+    de = rd.docket_entry
+    d = de.docket
+    court = d.court
```
FWIW, we're moving away from async in general. There's a good chance we'll be yanking it just about everywhere.

florean replied:
@elisa-a-v, correct, Django will load those and I'm avoiding the three DB call that were there because I know they're already loaded.  But I can add a comment reminding people calling it could result in DB loads.

florean replied:
I've added a comment.

florean commented:
### 1. Docs with many attachments
> For example I found [this item](https://www.courtlistener.com/docket/17366820/akai-custom-guns-llc-v-kkm-precision-inc/?entry_gte=207#entry-207) with 212 attachments, and it overflows the page:

Yeah, the UI needs to be able to deal with the worse case, even if it's a non-interactive "...and 150 more attachments" or something.  I'll work on that.

> Hiding the dropdown for docs with too many attachments seems like the easy way out here

Unfortunately, at that point the query is done.  How much of a problem is it?  My guess is that is an extreme example, but I don't know the dataset and I don't know what reasonable targets are.  The questions I can't answer are: what's the median, 95th, and 99th percentile for docket entry documents (technically you'd want to calculate it off of actual page views, but this is sufficient).   Could you run this for me on whatever would have a representative dataset?
```SQL
SELECT
	PERCENTILE_CONT(.5) WITHIN GROUP (ORDER BY entry_count) AS median,
	PERCENTILE_CONT(.95) WITHIN GROUP (ORDER BY entry_count) AS p95,
	PERCENTILE_CONT(.99) WITHIN GROUP (ORDER BY entry_count) AS p99
FROM
	(SELECT COUNT(*) AS entry_count FROM search_recapdocument GROUP BY docket_entry_id) AS _ 
;
```
Because I think you'd take the trade if the 95th percentile page load drops, but the 99th goes up a little.  If this a problem, I think what I would do is:
• only get the pk and attachment number in the first query, which is all that's needed for the first few checks.
• get the full RECAPDocument for the correct one.
• get only what I need for the other attachments.
I don't know, I might try that anyway, I might be able to make things a little faster even in the common case.

> Tip on data cloning for local testing

Thanks, I didn't have API access at the time and hadn't messed with factories, so I was generating everything in the admin.  It was painful.

### 2. Other RECAPDocument tabs
> I only see the dropdown button when in the default PDF/Text pages, but it disappears when in the authorities tab

I didn't test any docs with authorities.  I'll load the one from your screenshot.  There is a completely different view for them, `view_recap_authorities`, so is it a different page load?  I'll take a look.  Glancing at the view, at a minimum, there is a lot of common logic that should be factored out.

florean commented:
Yeah, so the Authorities tab is a page load to `.../authorities`, which is a completely different view and template.  Well, I say completely different, but really they're copies.  It is basically a very roundabout way to avoid making [this query](https://github.com/freelawproject/courtlistener/blob/main/cl/search/models.py#L3499) unless absolutely necessary.  And yeah, that doesn't look like a query you want to be making if you don't have to, but isn't this something we could load dynamically with HTMX?

mlissner commented:
What we do on the docket page (which has similar design of metadata at the top and tabs a few inches down) is have a common template for the top stuff and then the tabs all just use it. I imagine that'd work well here? I don't think moving to HTMX for that is the best plan just because it's already working and we don't usually use it for the main page contents (we use it more for little things like the attachment page, typically). 

There is a process for generating authorities, but I don't know what it is anymore. It's run automatically when we ingest data, so I suspect it didn't run because adding from admin doesn't trigger that code path. The function will be something like `find_citations...`, if I recall correctly. You could also just fake the data, I imagine.

I'll let Elisa help with the rest. :)

mlissner commented on cl/opinion_page/views.py:
```diff
@@ -577,7 +577,7 @@ async def make_thumb_if_needed(
             max_dimension=1068,
         )
-        await rd.arefresh_from_db()
+        await rd.arefresh_from_db(fields=["thumbnail_status", "thumbnail"])
```
I agree, that sounds best.

elisa-a-v reviewed (commented):
I haven't performed a thorough review of the latest changes but I wanted to share two small comments before you finish!

elisa-a-v commented:
Bah, didn't notice you had already finished 🤦🏽 my apologies! I'll provide a full review tomorrow, the comments above are small and unrelated to your latest changes so I think it should be fine.

florean commented on cl/opinion_page/urls.py:
Yes, that's a bug. I was matching the view's type hint, but the model is what matters. I'll remove it.

florean commented:
> Bah, didn't notice you had already finished 🤦🏽 my apologies!

They're small changes and I'll push them when I have a sec.

florean commented on cl/opinion_page/views.py:
```diff
@@ -679,8 +742,16 @@ async def view_recap_document(
-    rd = await make_thumb_if_needed(request, rd)
+    title = make_rd_title(rd)
+    needs_thumb = rd.thumbnail_status != THUMBNAIL_STATUSES.COMPLETE
+    if all([needs_thumb, rd.has_valid_pdf, is_og_bot(request)]):
```
So we have `is_og_bot()` but we also get passed in an `is_og_bot` bool.  Can I believe the passed in value?  And if not, should it be removed and the call be moved further up?  Right now we use both.

mlissner replied:
Yeah, I think you're right. We can just trust what comes in and don't need to run the `is_og_bot` function.

elisa-a-v commented:
@florean are you planning on pushing any more changes? it's perfectly fine if you are, but in that case could you please mark as draft and only mark as ready for review once all changes have been pushed? 🙏🏽

mlissner commented on cl/opinion_page/views.py:
```diff
@@ -613,18 +616,70 @@ async def view_recap_document(
+        is_og_bot,
+        "recap_authorities.html",
+    )
+    if isinstance(response, SimpleTemplateResponse):
```
Just want to flag that we had an issue with this template type: https://github.com/freelawproject/courtlistener/pull/5962

florean commented:
@elisa-a-v No, it's done.  I had to force everything after my failed merge, but they were ones that were already there and I just ran a (correct) merge.  There haven't been any code changes since last night.

Instead of merging `upstream/main` because I have a dirty main in order to test templates, I merged `upstream main`.  🤦‍♂️

mlissner reviewed (approved):
I didn't read in great depth, but this one feels fully cooked. Going to trust Elisa's earlier review and Nathan's efforts and merge. I'll revert if I should have looked more closely.
